### PR TITLE
fix(roles): add missing filter scope for surveys and static contents

### PR DESCRIPTION
**Pull request by donni106**
- added `filtered_for_current_user` when creating or updating surveys and static contents to respect roles entirely
- added `compact` on retrieving `survey_poll_ids` to avoid crashes when `dateable` map returns an array with `nil`s
- updated scope for static contents to restrict abilities just for admins and editors

SVA-388

### CODE REVIEW

File: app/graphql/mutations/create_or_update_static_content.rb [hunk 1]
```diff
@@ -36,6 +37,8 @@ def create_or_update(params)
       end
 
       def resource_or_error_message(record)
+        return {} if record.blank?
+
```
Comment (tonekk) [inline app/graphql/mutations/create_or_update_static_content.rb:41]: Would suggest implementing this whole section as follows:

```ruby
      class UnauthorizedError < StandardError
        def initialize(user)
          super("#{user.email} is not authorized to edit static contents")
        end
      end

      def create_or_update(params)
        static_content_id = params.delete(:id)
        user = context[:current_user]

        begin
          if static_content_id.present?
            static_content = StaticContent.find(static_content_id)
            raise UnauthorizedError.new(user) unless user.admin_role? || user.editor_role?
            static_content&.update(params)
          else
            static_content = StaticContent.create!(params)
          end
        rescue ActiveRecord::RecordInvalid, ActiveRecord::RecordNotFound, UnauthorizedError => e
          raise GraphQL::ExecutionError.new(e.message)
        end

        static_content
      end
```

File: app/models/static_content.rb [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
-  # we need the scope for destroying records, but we do not have any specifics
-  # for current users at static contents, so we return `all`
-  scope :filtered_for_current_user, lambda { |_current_user|
-    all
+  # we need the scope for destroying records, which only admins and editors are allowed to
+  scope :filtered_for_current_user, lambda { |current_user|
+    return all if current_user.admin_role?
+    return all if current_user.editor_role?
+
+    none
```
Comment (tonekk) [inline app/models/static_content.rb:9]: I think this is fairly misleading

File: app/graphql/mutations/create_or_update_static_content.rb [hunk 3]
```diff
@@ -26,8 +26,9 @@ def create_or_update(params)
         static_content_id = params.delete(:id)
```
[the remark this answers is not in the document]
Comment (donni106) [reply app/graphql/mutations/create_or_update_static_content.rb:26]: i think to provide some kind of backward compatibility to some services that call this one mutation and we cannot update everywhere at the same time. but i am not sure 100%.

File: app/models/static_content.rb (hunk 2 above)
Comment (donni106) [reply app/models/static_content.rb:9]: why? we need an active record relation returned here, but without any data. what would be your different approach?

File: app/graphql/mutations/create_or_update_static_content.rb [hunk 4]
```diff
@@ -36,6 +37,8 @@ def create_or_update(params)
       end
 
       def resource_or_error_message(record)
+        return {} if record.blank?
```
[the remark this answers is not in the document]
Comment (donni106) [reply app/graphql/mutations/create_or_update_static_content.rb:40]: if we return an error here, for example `return GraphQL::ExecutionError.new("Some Error") if record.blank?` then the outcome would be:

```
{
  "data": null,
  "errors": [
    {
      "message": "Some Error",
      "locations": [
        {
          "line": 7,
          "column": 2
        }
      ],
      "path": [
        "createOrUpdateStaticContent"
      ]
    }
  ]
}
```

compared to:

```
{
  "data": {
    "createOrUpdateStaticContent": {
      "id": null,
      "name": null,
      "content": null,
      "dataType": null
    }
  }
}
```

with returning the empty hash. clients can access the expected data this way and see, that the requested something did not return matching data. this is not an error of the graphql or something. its "just" returning no data for example for an id that does not exist. i think this shouldn't be a case of "crashing" the communication between client and server?

File: app/graphql/mutations/create_or_update_static_content.rb [hunk 5]
```diff
@@ -26,8 +26,9 @@ def create_or_update(params)
         static_content_id = params.delete(:id)
 
         if static_content_id.present?
-          static_content = StaticContent.find(static_content_id)
-          static_content.update(params)
+          static_content = StaticContent.filtered_for_current_user(context[:current_user])
```
[the remark this answers is not in the document]
Comment (donni106) [reply app/graphql/mutations/create_or_update_static_content.rb:29]: i see your point, but `filtered_for_current_user` is widely used across our resources. so it is here for consistency. rethinking "the whole thing" is out of scope but we can put it on the roadmap for sure.

File: app/models/static_content.rb (hunk 2 above)
Comment (tonekk) [reply app/models/static_content.rb:9]: Misleading, not wrong.
It doesn't "really" filter them for the `current_user`, I wrote an (in my opinion) better solution above.

File: app/graphql/mutations/create_or_update_static_content.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (tonekk) [reply app/graphql/mutations/create_or_update_static_content.rb:40]: This is not crashing the communication between client and server.
I think it's fair expect some error handling on the client side - just like the case of an "empty" response has to be handled.

I'm not sure what's the best practice here. Personally I would go with throwing errors.
[The apollo wiki lists error codes that speak for my approach](https://www.apollographql.com/docs/apollo-server/data/errors/).
[This blog posts suggests to use "ResultTypes" to differentiate results from "real" errors.](https://sachee.medium.com/200-ok-error-handling-in-graphql-7ec869aec9bc) Not sure if I would apply this example for a request that is unauthorized, though.

File: app/graphql/mutations/create_or_update_static_content.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (tonekk) [reply app/graphql/mutations/create_or_update_static_content.rb:29]: I see.

File: app/models/static_content.rb (hunk 2 above)
Comment (donni106) [reply app/models/static_content.rb:9]: i understand. but we also need the scope for `Mutations::DestroyRecord` for example. so refactoring with your suggestion will lead to a more complex job across multiple places.

**Review by marcometz — APPROVED** [review]
Das Thema ErrorHandling ist wichtig, aber ich befürchte das sprengt hier gerade etwas den Rahmen und wir müssten ausprobieren ob auch die Native-APPs mit einer veränderten Antwort klar kommen wurden.

Ich hätte gerne, dass wir das separat angehen als eigenständiges Thema. @tonekk Kannst du dir vorstellen hierzu einen PR mit Test vorzubereiten? 

Vielen Dank an euch beide!

**Comment by tonekk** [comment]
Yo, kann ich machen

**Review by tonekk — APPROVED** [review]

**Pull request merged**